src/main/java/dev/jbang/dependencies/ArtifactResolver.java

```diff
@@ -0,0 +1,459 @@
+																	.addService(TransporterFactory.class,
+																			HttpTransporterFactory.class);
+		return locator.getService(RepositorySystem.class);
+	}
+
+	private Settings newEffectiveSettings() {
+		DefaultSettingsBuilderFactory factory = new DefaultSettingsBuilderFactory();
+		DefaultSettingsBuilder builder = factory.newInstance();
+
+		SettingsBuildingRequest settingsBuilderRequest = new DefaultSettingsBuildingRequest();
+		settingsBuilderRequest.setSystemProperties(System.getProperties());
+
+		// if we have a root folder, don't read settings at all
+		if (withUserSettings && (localFolderOverride == null || settingsXml != null)) {
+			// find the settings
+			Path settingsFile = settingsXml;
+			if (settingsFile == null) {
+				Path userSettings = Paths.get(System.getProperty("user.home"), ".m2", "settings.xml");
+				if (Files.exists(userSettings)) {
+					settingsFile = userSettings.toAbsolutePath();
+				}
+			}
+			if (settingsFile != null) {
+				settingsBuilderRequest.setUserSettingsFile(settingsFile.toFile());
+			}
+		}
+
+		// read it
+		SettingsBuildingResult settingsBuildingResult;
+		try {
+			settingsBuildingResult = builder.build(settingsBuilderRequest);
+		} catch (SettingsBuildingException e) {
+			throw new RuntimeException(e);
+		}
+
+		return settingsBuildingResult.getEffectiveSettings();
+	}
+
+	private DefaultRepositorySystemSession newSession(RepositorySystem system, Settings settings) {
+		DefaultRepositorySystemSession session = MavenRepositorySystemUtils.newSession();
+		LocalRepository localRepo = newLocalRepository(settings);
+		session.setLocalRepositoryManager(system.newLocalRepositoryManager(session, localRepo));
+		return session;
+	}
+
+	private LocalRepository newLocalRepository(Settings set) {
+		String localRepository = localFolderOverride != null ? localFolderOverride.toString() : null;
+		if (localRepository == null && set != null)
+			localRepository = set.getLocalRepository();
+		if (localRepository == null)
+			localRepository = System.getProperty("user.home") + File.separator + ".m2" + File.separator + "repository";
+		else if (!new File(localRepository).isAbsolute())
+			localRepository = Util.getCwd().resolve(localRepository).toAbsolutePath().toString();
+		return new LocalRepository(localRepository);
+	}
+
+	private void configureSession(DefaultRepositorySystemSession session, Settings settings) {
+		// set up authentication
+		DefaultAuthenticationSelector authenticationSelector = new DefaultAuthenticationSelector();
+		for (Server server : settings.getServers()) {
+			AuthenticationBuilder auth = new AuthenticationBuilder();
+			if (server.getUsername() != null)
+				auth.addUsername(server.getUsername());
+			if (server.getPassword() != null)
+				auth.addPassword(server.getPassword());
+			if (server.getPrivateKey() != null)
+				auth.addPrivateKey(server.getPrivateKey(), server.getPassphrase());
+			authenticationSelector.add(server.getId(), auth.build());
```

Comment:
For custom headers, which for example gitlab requires, there still needs to be some processing done

```
if (server.getConfiguration() != null) {
	Xpp3Dom dom = (Xpp3Dom) server.getConfiguration();
	for (int i = dom.getChildCount() - 1; i >= 0; i--) {
		Xpp3Dom child = dom.getChild(i);
		if ("httpHeaders".equals(child.getName())) {
			HashMap<String, String> headers = new HashMap<>();
			Xpp3Dom[] properties = child.getChildren("property");
			for (Xpp3Dom property : properties) {
				headers.put(property.getChild("name").getValue(), property.getChild("value").getValue());
			}
			session.setConfigProperty(ConfigurationProperties.HTTP_HEADERS + "." + server.getId(), headers);
		}
	}
}
```

This will configure the aether http connector to send these headers

Reply:
@JakobFels [here](https://github.com/jbangdev/jbang/issues/1077#issuecomment-1237721810) you commented with a [link to code](https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java#L238-L263) that's very different from this. Do you know which one is the correct code? (And out of curiosity, where did you get this new code from?)

Reply:
@JakobFels can you try the new build? It uses the code you posted above. Let's see if that works...

Reply:
With these additions jbang could resolve an artifact hosted at a gitlab maven registry.

> @JakobFels [here](https://github.com/jbangdev/jbang/issues/1077#issuecomment-1237721810) you commented with a [link to code](https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java#L238-L263) that's very different from this. Do you know which one is the correct code? (And out of curiosity, where did you get this new code from?)

I stepped through parts of the code to find out where the headers are read from config.
The relevant code is [here](https://github.com/apache/maven-resolver/blob/master/maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java#L153).
I then tried the linked code from before, until I noticed that the properties are different (aether.connector.wagon vs aether.connector.http).
The current result was my attempt to read these settings.

The first link was me poking at maven code to see where maven itself pulls these settings from, which seems to have been the wrong place (I still don't quite understand when maven sets these up.)

Reply:
> I still don't quite understand when maven sets these up

yeah documentation is severely lacking, so it's down to looking at the code.

Btw, will you have time to look at the latest build to see if it works for you?

Reply:
> Btw, will you have time to look at the latest build to see if it works for you?

The current build works for my setup, artifacts hosted on gitlab resolve properly.